### ISSUE TITLE
io::RealtimeExecutor breaks boost::asio::post() semantics

### ISSUE THREAD

**Issue by bmartin427** [mjbots/mjlib#3]
`boost::asio::post()` has the following [documentation](https://www.boost.org/doc/libs/1_74_0/doc/html/boost_asio/reference/post/overload2.html):

> The function object is queued for execution, and **is never called from the current thread prior to returning from post()**.

However, somehow this does not seem to be the case when operating through an `mjlib::io::RealtimeExecutor`.  To demonstrate, add the following code to the end of the test case in `realtime_executor_test.cc`:

```
  boost::asio::any_io_executor executor{dut};

  bool cb1_called{};
  bool cb2_called{};
  const auto cb2 = [&]() {
    BOOST_TEST(cb1_called);
    BOOST_TEST(!cb2_called);
    cb2_called = true;
  };
  const auto cb1 = [&]() {
    BOOST_TEST(!cb1_called);
    BOOST_TEST(!cb2_called);
    cb1_called = true;
    boost::asio::post(executor, cb2);
    BOOST_TEST(!cb2_called);
  };
  boost::asio::post(executor, cb1);
  BOOST_TEST(!cb1_called);
  BOOST_TEST(!cb2_called);
  context.poll();
  context.restart();
  BOOST_TEST(cb1_called);
  BOOST_TEST(cb2_called);
```

The check that `cb2_called` is false at the end of `cb1` fails.  If the `any_io_executor` used is constructed with `context.get_executor()` instead of `dut`, it passes.

I don't yet understand exactly how `RealtimeExecutor` is going wrong.  I do know that the call stack from `post()` goes through `RealtimeExecutor::execute()` directly to the callback that should have been deferred.  I haven't found any good documentation for what `execute()` is supposed to do though, so I'm unclear if `RealtimeExecutor` should have done something other than call `base_.execute()`, or if something has already gone wrong prior to that.

**Comment by jpieper** [comment]
I agree this seems broken.  The documentation describing the semantic differences between an executor and an io_context is a shambles, as this was likely an intermediate state between various updatings of ASIO to match the evolving C++ committees plan for asynchronous operation.

I honestly haven't touched this or used it in 4 years and I can't say I made much progress looking at it for 2 hours here.  What is the urgency / your desire to fix it / your ability to switch to something else?

**Comment by jpieper** [comment]
Actually, I am able to get the test to pass with the following pre-amble:

```
  boost::asio::io_context context;
  auto ex1 = context.get_executor();
  auto ex2 = boost::asio::require(ex1,
                                  boost::asio::execution::blocking.never);
  RealtimeExecutor dut{ex2};

  boost::asio::any_io_executor executor{dut};
```

That may be enough to make progress...?

What I did learn was that boost::post seems to invoke `execute` on the `any_io_executor`.  That in turn relies on the "possibly_blocking" configurable parameter, which is apparently either not set up, or not forwarded correctly.  Almost certainly my problem from trying to port this to the executor world when asio moved there, but AFAIK the C++ world has moved well beyond that, so I'm not super inclined to spend a lot of time trying to figure out how to make it work in this intermediate state.

**Comment by bmartin427** [comment]
I think I have it fixed now actually.  As you showed, it's related to the property system; i.e. this test case fails:
```
  boost::asio::io_context context;
  RealtimeExecutor dut{context.get_executor()};
  boost::asio::any_io_executor executor{dut};

  EXPECT_EQ(boost::asio::query(
                boost::asio::require(executor, boost::asio::execution::blocking.never),
                boost::asio::execution::blocking),
            boost::asio::execution::blocking.never);
```

I was able to fix this by changing `Base` to be `boost::asio::any_io_executor`, changing `require()` to `return RealtimeExecutor(boost::asio::require(base_, value))`, changing the other wrapped methods to go through the `boost::asio::` free functions instead of invoking methods on the executor, and deleting the work started/finished private methods which seemed to be unused.

I don't actually need this fixed upstream, as I have my own copy I've applied the fixes to, I just thought it might be useful to report my findings.  Thanks for looking! :smile:
